### PR TITLE
feat: CXSPA-1519 Refine display of value prices (absolute/delta)

### PR DESCRIPTION
**Pull request by Stofftiger**
- Display selected Value as "selected"
 - Display prices for currently not selected DDLB values
 - fixed old tests and added new tests

### CODE REVIEW

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts [hunk 1]
```diff
@@ -31,7 +31,14 @@ export class ConfiguratorPriceComponent {
     if (this.formula.priceTotal) {
       return this.priceTotal;
     } else {
-      return '+ ' + this.formula.price?.formattedValue;
+      if (
+        this.formula.price?.value !== undefined &&
+        this.formula.price?.value >= 0
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: You could simply write:
`this.formula.price?.value >= 0`
This statement is not necessary `this.formula.price?.value !== undefined` because with` this.formula.price?` will be checked  whether ` this.formula.price `is defined

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts [hunk 2]
```diff
@@ -41,7 +48,14 @@ export class ConfiguratorPriceComponent {
    * @return {string} - total price formula
    */
   get priceTotal(): string {
-    return '+ ' + this.formula.priceTotal?.formattedValue;
+    if (
+      this.formula.priceTotal !== undefined &&
+      this.formula.priceTotal.value >= 0
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: The same here:
You could simply write:
`this.formula.priceTotal?.value >= 0`
This statement is not necessary `this.formula.priceTotal !== undefined` because with` this.formula.priceTotal?` will be checked  whether ` this.formula.priceTotal `is defined

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts [hunk 3]
```diff
@@ -31,7 +31,14 @@ export class ConfiguratorPriceComponent {
...
       return this.priceTotal;
     } else {
-      return '+ ' + this.formula.price?.formattedValue;
+      if (
+        this.formula.price?.value !== undefined &&
+        this.formula.price?.value >= 0
+      ) {
+        return '+' + this.formula.price?.formattedValue;
+      } else {
+        return this.formula.price?.formattedValue || '';
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: Plz, explain what you expected to get here. 
`this.formula.price?.formattedValue || ''` syntax is not known.
According to debugger `this.formula.price?.formattedValue` will be returned.

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts [hunk 4]
```diff
@@ -41,7 +48,14 @@ export class ConfiguratorPriceComponent {
...
    */
   get priceTotal(): string {
-    return '+ ' + this.formula.priceTotal?.formattedValue;
+    if (
+      this.formula.priceTotal !== undefined &&
+      this.formula.priceTotal.value >= 0
+    ) {
+      return '+' + this.formula.priceTotal?.formattedValue;
+    } else {
+      return this.formula.priceTotal?.formattedValue || '';
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts:55]: Plz, explain what you expected to get here. 
`this.formula.priceTotal?.formattedValue || ''` syntax is not known.
According to debugger `this.formula.priceTotal?.formattedValue` will be returned.

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts [hunk 5]
```diff
@@ -54,6 +68,14 @@ export class ConfiguratorPriceComponent {
     return quantity ? this.formula.price?.value !== 0 && quantity >= 1 : false;
   }
 
+  /**
+   * Verifies if 'Selected' should be shown
+   *
+   * @return {boolean} - 'true' if 'Selected' should be shown, otherwise 'false'
+   */
+  displaySelected(): boolean {
+    return this.isPriceLightedUp() && !this.displayFormula();
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: The wording `Selected` should be the same colour as price:
image

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: The same as in Accelerator:
image

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts [hunk 6]
```diff
@@ -146,16 +146,32 @@ export class ConfiguratorAttributeBaseComponent {
...
+    configuratorValue?: Configurator.Value | undefined
   ): string {
     let title = label ? label : '';
     if (expMode && techName) {
       title += ` / [${techName}]`;
     }
+    if (configuratorValue && !configuratorValue.selected) {
+      if (
+        configuratorValue.valuePrice?.value !== undefined &&
+        configuratorValue.valuePrice.value < 0
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: If you add `?` after `configuratorValue` and `valuePrice` then `configuratorValue?.valuePrice?.value` then `configuratorValue.valuePrice?.value !== undefined` is not neccessary

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts [hunk 7]
```diff
@@ -146,16 +146,32 @@ export class ConfiguratorAttributeBaseComponent {
...
    * @param techName - value technical name
+   * @param configuratorValue - item Object which contains the valuePrice
+   * @param isSelected - is item in dropdownlist selected
    */
   getLabel(
     expMode: boolean,
     label: string | undefined,
-    techName: string | undefined
+    techName: string | undefined,
+    configuratorValue?: Configurator.Value | undefined
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: Plz,  rename to `value`

Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]:
```suggestion
    configuratorValue?: Configurator.Value
```

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts [hunk 8]
```diff
@@ -146,16 +146,32 @@ export class ConfiguratorAttributeBaseComponent {
...
       title += ` / [${techName}]`;
     }
+    if (configuratorValue && !configuratorValue.selected) {
+      if (
+        configuratorValue.valuePrice?.value !== undefined &&
+        configuratorValue.valuePrice.value < 0
+      ) {
+        title += ` [${configuratorValue.valuePrice?.formattedValue}]`;
+      } else if (
+        configuratorValue.valuePrice?.value !== undefined &&
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: If you add `?` after `configuratorValue` and `valuePrice` then `configuratorValue?.valuePrice?.value` then `configuratorValue.valuePrice?.value !== undefined` is not neccessary

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts [hunk 9]
```diff
@@ -146,16 +146,32 @@ export class ConfiguratorAttributeBaseComponent {
    * @param expMode - Is expert mode set?
    * @param label - value label
    * @param techName - value technical name
+   * @param configuratorValue - item Object which contains the valuePrice
+   * @param isSelected - is item in dropdownlist selected
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: Typi: dropdown list

File: feature-libs/product-configurator/rulebased/components/attribute/types/drop-down/configurator-attribute-drop-down.component.html [hunk 10]
```diff
@@ -24,12 +24,12 @@
...
     <option
       *ngFor="let item of attribute.values"
-      [label]="getLabel(expMode, item.valueDisplay, item.valueCode)"
       [selected]="item.selected"
+      [label]="getLabel(expMode, item.valueDisplay, item.valueCode, item)"
       [attr.aria-label]="getAriaLabel(item, attribute)"
       [value]="item.valueCode"
     >
-      {{ getLabel(expMode, item.valueDisplay, item.valueCode) }}
+      {{ getLabel(expMode, item.valueDisplay, item.valueCode, item) }}
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/drop-down/configurator-attribute-drop-down.component.html:32]: There is an issue with a dropdown list. For `Please select` there is a price:
image

File: feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.html [hunk 11]
```diff
@@ -1,7 +1,4 @@
-<div
-  id="{{ createAttributeIdForConfigurator(attribute) }}"
-  *ngIf="attribute.values !== undefined && attribute.values.length > 0"
->
+<div class="form-group" id="{{ createAttributeIdForConfigurator(attribute) }}">
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.html]: This is breaking and needs to be equipped with the cxFeatureLevel directive

File: feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.html [hunk 12]
```diff
@@ -24,10 +21,21 @@
                 }
         }}
       </span>
-      <div class="cx-read-only-label" *ngIf="value.selected" aria-hidden="true">
-        <span aria-hidden="true">{{
-          getLabel(expMode, value.valueDisplay, value.valueCode)
-        }}</span>
+      <div *ngIf="value.selected" aria-hidden="true">
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.html]: Also breaking

File: feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.spec.ts [hunk 13]
```diff
@@ -57,27 +74,32 @@ describe('ConfigAttributeReadOnlyComponent', () => {
     expect(component).toBeTruthy();
   });
 
+  it('should test return of extractValuePriceFormulaParameters() method', () => {
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.spec.ts]: Please rephrase (the method should do something...)

File: feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.ts [hunk 14]
```diff
@@ -16,4 +17,21 @@ export class ConfiguratorAttributeReadOnlyComponent extends ConfiguratorAttribut
...
+  /**
+   * Extract corresponding value price formula parameters.
+   * For the multi-selection attribute types the complete price formula should be displayed at the value level.
+   *
+   * @param {Configurator.Value} value - Configurator value
+   * @return {ConfiguratorPriceComponentOptions} - New price formula
+   */
+  extractValuePriceFormulaParameters(
+    value: Configurator.Value
+  ): ConfiguratorPriceComponentOptions | undefined {
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.ts]:
```suggestion
  ): ConfiguratorPriceComponentOptions  {
```

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.html [hunk 15]
```diff
@@ -17,3 +17,8 @@
     <div class="cx-price-total">{{ priceTotal }}</div>
   </ng-container>
 </ng-container>
+<ng-container *ngIf="displaySelected()">
+  <div [ngClass]="styleClass">
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.html]: Breaking, needs to be handled

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts [hunk 16]
```diff
@@ -97,7 +97,58 @@ describe('ConfiguratorPriceComponent', () => {
         '.cx-quantity-price'
       );
 
-      expect(component.price).toEqual('+ $10');
+      expect(component.price).toEqual('+$10');
+    });
+
+    it('should be defined: value price less than zero', () => {
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: What is that supposed to mean? Please rephrase
(I understand that other tests in this file also have similar expressions)

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts [hunk 17]
```diff
@@ -97,7 +97,58 @@ describe('ConfiguratorPriceComponent', () => {
...
+      );
+      CommonConfiguratorTestUtilsService.expectElementNotPresent(
+        expect,
+        htmlElem,
+        '.cx-quantity-price'
+      );
+
+      expect(component.price).toEqual('-$10');
+    });
+    it('should be defined: value price is undefined', () => {
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: See above

File: feature-libs/product-configurator/rulebased/styles/_configurator-attribute-read-only.scss [hunk 18]
```diff
@@ -1,5 +1,6 @@
 %cx-configurator-attribute-read-only {
   @include cx-configurator-attribute-type();
+  @include cx-configurator-form-group();
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/styles/_configurator-attribute-read-only.scss]: Breaking change, needs to be equipped with feature flag. See https://sap.github.io/spartacus-docs/css-architecture/#style-versioning

**Review by ChristophHi — CHANGES_REQUESTED** [review]

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts (hunk 7 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: done

Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: done

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts (hunk 6 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: done

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts (hunk 8 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: done

File: feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.spec.ts (hunk 13 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.spec.ts]: done

File: feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.ts (hunk 14 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/read-only/configurator-attribute-read-only.component.ts]: done

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts (hunk 1 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: done

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts (hunk 3 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: '' + this.formula.price?.formattedValue || '' will return a String containing "undefined" 
this.formula.price?.formattedValue || '' will return a String of length 0: '' when the first term is undefined.
since the return type of the method is String its needed to make sure always a String with the formattedValue or length 0 is returned.

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts (hunk 2 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: done

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts (hunk 4 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts:55]: See above

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts (hunk 17 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: done

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts (hunk 5 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: done

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts [hunk 19]
```diff
@@ -146,16 +146,26 @@ export class ConfiguratorAttributeBaseComponent {
    * @param expMode - Is expert mode set?
    * @param label - value label
    * @param techName - value technical name
+   * @param value - item Object which contains the valuePrice
+   * @param isSelected - is item in dropdown list selected
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: Adjust your comment, there is no `isSelected` parameter

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts [hunk 20]
```diff
@@ -146,16 +146,26 @@ export class ConfiguratorAttributeBaseComponent {
    * @param expMode - Is expert mode set?
    * @param label - value label
    * @param techName - value technical name
+   * @param value - item Object which contains the valuePrice
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: By the way a value contains another items

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts [hunk 21]
```diff
@@ -146,16 +146,26 @@ export class ConfiguratorAttributeBaseComponent {
...
     label: string | undefined,
-    techName: string | undefined
+    techName: string | undefined,
+    value?: Configurator.Value
   ): string {
     let title = label ? label : '';
     if (expMode && techName) {
       title += ` / [${techName}]`;
     }
+    if (value && !value.selected) {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: Is mor readable:

> if (!value.selected && value.valuePrice?.value) {
>       if (value.valuePrice.value < 0) {
>         title += ` [${value.valuePrice.formattedValue}]`;
>       } else if(value.valuePrice.value >= 1) {
>         title += ` [+${value.valuePrice.formattedValue}]`;
>       }
>     }

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: The tests are still running green

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: Plz, extract into separate method, for instance `protected getValuePrice(title: string, value: Configurator.Value | undefined)`

Then you could simple call the previously defined method, like:

> return value ? this.getValuePrice(title, value) : title;

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts [hunk 22]
```diff
@@ -182,5 +182,37 @@ describe('ConfigUIKeyGeneratorService', () => {
       const techLabel = label + ' / [' + techName + ']';
       expect(classUnderTest.getLabel(true, label, techName)).toEqual(techLabel);
     });
+    it('should return label without technical name but with positive price', () => {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts]: Plz, remove `without technical name but`. It is irrelevant for your purposes.
The goal of the test is to test whether the positive price will be returned.

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts [hunk 23]
```diff
@@ -182,5 +182,37 @@ describe('ConfigUIKeyGeneratorService', () => {
...
+      };
+      const configuratorValue: Configurator.Value = {
+        valueCode: 'valueCode',
+        valuePrice: priceDetails,
+      };
+      expect(
+        classUnderTest.getLabel(false, label, techname, configuratorValue)
+      ).toEqual('label [+$3.20]');
+    });
+    it('should return label without technical name but with negative price', () => {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts]: The same here: Plz, remove `without technical name but`. It is irrelevant for your purposes.
The goal of the test is to test whether the negative price will be returned.

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts [hunk 24]
```diff
@@ -182,5 +182,37 @@ describe('ConfigUIKeyGeneratorService', () => {
       const techLabel = label + ' / [' + techName + ']';
       expect(classUnderTest.getLabel(true, label, techName)).toEqual(techLabel);
     });
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts:189]: Plz, add a test for zero price. It expects `label` as return value and not `label [+$0.0]`. 
The end-user wants to see price when it is cheaper or more expensive but not when the price equals zero.

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts [hunk 25]
```diff
@@ -182,5 +182,37 @@ describe('ConfigUIKeyGeneratorService', () => {
...
+        value: 3.2,
+      };
+      const configuratorValue: Configurator.Value = {
+        valueCode: 'valueCode',
+        valuePrice: priceDetails,
+      };
+      expect(
+        classUnderTest.getLabel(false, label, techname, configuratorValue)
+      ).toEqual('label [+$3.20]');
+    });
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts:198]: Plz, add empty line for readability

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts (hunk 22 above)
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts]: Plz, add empty line for readability

File: feature-libs/product-configurator/rulebased/components/attribute/types/drop-down/configurator-attribute-drop-down.component.html (hunk 10 above)
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/attribute/types/drop-down/configurator-attribute-drop-down.component.html:32]: Plz, check with our a11y expert whether value are read correctly with and without price

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts [hunk 26]
```diff
@@ -41,7 +45,11 @@ export class ConfiguratorPriceComponent {
    * @return {string} - total price formula
    */
   get priceTotal(): string {
-    return '+ ' + this.formula.priceTotal?.formattedValue;
+    if (this.formula.priceTotal && this.formula.priceTotal.value >= 0) {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts:52]: I think the condition should be `this.formula.priceTotal.value >= 1` and not `this.formula.priceTotal.value >= 0`.
The end-user wants to see price when it is cheaper or more expensive but not when the price equals zero.
Plz, add a test for zero case.

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts [hunk 27]
```diff
@@ -31,7 +31,11 @@ export class ConfiguratorPriceComponent {
     if (this.formula.priceTotal) {
       return this.priceTotal;
     } else {
-      return '+ ' + this.formula.price?.formattedValue;
+      if (this.formula.price?.value && this.formula.price?.value >= 0) {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: I think the condition should be `this.formula.priceTotal.value >= 1` and not `this.formula.priceTotal.value >= 0`.
The end-user wants to see price when it is cheaper or more expensive but not when the price equals zero.
Plz, add a test for zero case.

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts [hunk 28]
```diff
@@ -54,6 +62,14 @@ export class ConfiguratorPriceComponent {
     return quantity ? this.formula.price?.value !== 0 && quantity >= 1 : false;
   }
 
+  /**
+   * Verifies if 'Selected' should be shown
+   *
+   * @return {boolean} - 'true' if 'Selected' should be shown, otherwise 'false'
+   */
+  displaySelected(): boolean {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: Plz, remove this method and all tests. It is not required.

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.html (hunk 15 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.html]: Do we want to downport this fix to an previous versions?

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.html [hunk 29]
```diff
@@ -17,3 +17,8 @@
     <div class="cx-price-total">{{ priceTotal }}</div>
   </ng-container>
 </ng-container>
+<ng-container *ngIf="displaySelected()">
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.html]: I would suggest to define an else case and add a template for it, like

> <ng-container *ngIf="displayFormula(); else selectedCase">
>   <ng-container *ngIf="displayPriceOnly()">
>     <div
>       [ngClass]="styleClass"
>       [attr.aria-label]="'configurator.a11y.valueSurcharge' | cxTranslate"
>     >
>       {{ price }}
>     </div>
>   </ng-container>
>   <ng-container *ngIf="displayQuantityAndPrice()">
>     <div
>       class="cx-quantity-price"
>       [attr.aria-label]="'configurator.a11y.valueSurcharge' | cxTranslate"
>     >
>       {{ quantityWithPrice(formula?.quantity | cxNumeric) }}
>     </div>
>     <div class="cx-price-total">{{ priceTotal }}</div>
>   </ng-container>
> </ng-container>
> <ng-template #selectedCase>
>   <ng-container *ngIf="isPriceLightedUp()">
>     <div [ngClass]="styleClass">
>       {{ 'configurator.price.selectedItem' | cxTranslate }}
>     </div>
>   </ng-container>
> </ng-template>

Then `displaySelected` method could be completely removed.
By the way, this method has not been tested as all.

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.html]: Plz, check with a11y whether the `Selected` will be read by the screen reader. I strongly believe not because the `attr.aria--label` is not added. It should be like:
 
> <div [ngClass]="styleClass"
>          [attr.aria-label]="'configurator.price.selectedItem' | cxTranslate">
>       {{ 'configurator.price.selectedItem' | cxTranslate }}
>     </div>

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts [hunk 30]
```diff
@@ -89,45 +89,90 @@ describe('ConfiguratorPriceComponent', () => {
...
 
-      expect(component.price).toEqual('+ $10');
+      expect(component.price).toEqual('+$10');
     });
 
-    it('should be defined: value price greater than zero and value is selected', () => {
-      component.formula = createTestData(0, 10, 10, true);
+    it('should be defined: negative value price is shown correctly', () => {
+      component.formula = createTestData(0, -10, 0, false, true);
+      component.formula = {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: You try to define formula twice. Plz, remove the second one

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts [hunk 31]
```diff
@@ -89,45 +89,90 @@ describe('ConfiguratorPriceComponent', () => {
...
         htmlElem,
         '.cx-quantity-price'
       );
 
-      expect(component.price).toEqual('+ $10');
+      expect(component.price).toEqual('-$10');
     });
+    it('should be defined: no value price is given and a String of length 0 should be returned', () => {
+      component.formula = createTestData(0, 0, 0, false, true);
+      component.formula = {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: You try to define formula twice. Plz, remove the second one

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts [hunk 32]
```diff
@@ -89,45 +89,90 @@ describe('ConfiguratorPriceComponent', () => {
...
         expect,
         htmlElem,
         '.cx-quantity-price'
       );
 
-      expect(component.price).toEqual('+ $10');
+      expect(component.price).toEqual('-$10');
     });
+    it('should be defined: no value price is given and a String of length 0 should be returned', () => {
+      component.formula = createTestData(0, 0, 0, false, true);
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: You have to refactor `createTestData` method

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts [hunk 33]
```diff
@@ -138,35 +183,74 @@ describe('ConfiguratorPriceComponent', () => {
...
       );
-      expect(component.priceTotal).toEqual('+ $150');
+      expect(component.priceTotal).toEqual('+$150');
     });
 
-    it('should be defined: complete price formula', () => {
-      component.formula = createTestData(2, 10, 20);
+    it('should be defined: a negative total value price is given', () => {
+      component.formula = createTestData(0, 0, -150, true);
+      component.formula = {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: You try to define formula twice. Plz, remove the second one

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts [hunk 34]
```diff
@@ -138,35 +183,74 @@ describe('ConfiguratorPriceComponent', () => {
...
         expect,
         htmlElem,
-        '.cx-price'
+        '.cx-quantity-price'
+      );
+      expect(component.priceTotal).toEqual('-$150');
+    });
+    it('should be defined: no total value price is given and a String of length 0 should be returned', () => {
+      component.formula = createTestData(0, 0, 0, true);
+      component.formula = {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: You try to define formula twice. Plz, remove the second one

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts (hunk 19 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: done

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts (hunk 21 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.ts]: done, just made a few small adaptions: 

changed the order for : 
if (!value.selected && value.valuePrice?.value) {
to
 if (value?.valuePrice?.value && !value.selected) {

also had to make sure that value is not undefined by changing It to "value?" 
the order change was necessary since otherwise for the first statement value.selected value could be possibly undefined

also I changed the 
} else if(value.valuePrice.value >= 1) {
to
} else if(value.valuePrice.value >0) {
since we could also have a price of +$0.5

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts (hunk 22 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts]: done

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts (hunk 23 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts]: done

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts (hunk 25 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts:198]: done

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts (hunk 22 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts]: done

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts (hunk 26 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts:52]: shouldn't we then go for 
this.formula.priceTotal.value > 0 
to also catch prices like +$0.5 ?

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts (hunk 27 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: same as above, prefer to go for 
this.formula.priceTotal.value > 0

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts (hunk 28 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.ts]: done

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts (hunk 30 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: done

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts (hunk 31 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: done

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts (hunk 33 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: done

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts (hunk 34 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: done

File: feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts (hunk 24 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/attribute/types/base/configurator-attribute-base.component.spec.ts:189]: done

File: feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts (hunk 32 above)
Comment (Stofftiger) [reply feature-libs/product-configurator/rulebased/components/price/configurator-price.component.spec.ts]: done

**Review by ChristophHi — APPROVED** [review]

**Comment by ChristophHi** [comment]
This will probably never be merged, we just keep it as a reference
